pkg/reconciler/taskrun/validate_resources.go

```diff
@@ -122,6 +138,25 @@ func validateParams(ctx context.Context, paramSpecs []v1beta1.ParamSpec, params
 		return fmt.Errorf("param types don't match the user-specified type: %s", wrongTypeParamNames)
 	}
 
+	// At this point, we have checked against missing/extra params and param type alignment
+	// We now want to make sure all keys of object type params are provided in taskrun params.
+	var missingObjectKeysParamNames []string
+	for _, param := range params {
+		if _, ok := neededObjectParamsKeys[param.Name]; !ok {
+			// Ignore any missing params - this happens when extra params were
+			// passed to the task that aren't being used.
+			continue
+		}
+		miss := list.DiffLeft(neededObjectParamsKeys[param.Name], providedObjectParamsKeys[param.Name])
+		if len(miss) != 0 {
+			missingObjectKeysParamNames = append(missingObjectKeysParamNames, param.Name)
+		}
+	}
+
+	if len(missingObjectKeysParamNames) != 0 {
+		return fmt.Errorf("taskrun params miss keys for params %v defined in properties", missingObjectKeysParamNames)
+	}
+
```

Comment:
Can you make this and the above code as a separate function like: ValidateParamAgainstSchema(ps ParamSpec, p Param)? This way we can use it elsewhere.
Also wrap ValidateObjectAgainstSchema() against another sharable function such as ValidateObjectAgainstSchema(p PropertySpec, object ArraryOrString) This way, it can be used in Chains and Results in Pipelines

Reply:
> Can you make this and the above code as a separate function like: ValidateParamAgainstSchema(ps ParamSpec, p Param)? This way we can use it elsewhere.

sgtm! I'd use `properties` instead of `Schema` just to make sure we use terms consistently. 

> Also wrap ValidateObjectAgainstSchema() against another sharable function such as ValidateObjectAgainstSchema(p PropertySpec, object ArraryOrString)

can you explain this a bit? what kind of another helper function are you looking for?

Reply:
I can see that `ValidateParamAgainstSchema ` can be helpful when you try to validate params, but this is also a functionality that TaskRun results want to use for objects, so having a separate ValidateObjectAgainstSchema(p PropertySpec, object ArraryOrString) can give us another level of sharable code. Let me know if this clarify what I'm trying to say

Reply:
Can you clarify the difference between the two functions' signatures? Just trying to understand the needs and what level of isolation I should refactor.

For `ValidateParamAgainstProperties `, is it `ValidateParamAgainstProperties(paramSpecs []v1beta1.ParamSpec, params []v1beta1.Param)`? Just thinking whether to make this function parallel to the existing `validateParams` OR to utilizing `ValidateParamAgainstProperties` inside `validateParams` by first filtering out all object params and passing to `ValidateParamAgainstProperties`. 
And what is the second argument `object` of `ValidateObjectAgainstSchema`?

Reply:
ValidateParamAgainstProperties understanding is correct.
ValidateObjectAgainstSchema is like:
```
ValidateObjectAgainstSchema(p PropertySpec, object ArraryOrString)
```

Reply:
>  Just thinking whether to make this function parallel to the existing `validateParams` OR to utilizing `ValidateParamAgainstProperties` inside `validateParams` by first filtering out all object params and passing to `ValidateParamAgainstProperties`. And what is the second argument `object` of `ValidateObjectAgainstSchema`?

`ValidateParamAgainstProperties` is for later uses in Tekton Chains, and `ValidateObjectAgainstSchema` is for later usage in Results

Reply:
`PropertySpec` is the spec for a ***specific*** key/property of the object param and this spec just contains the type of the key, whereas `ArraryOrString` contains `objectVal` which is a map containing many key-value pairs. They are not comparable not can be validated against each other.

I extracted object-specific validation into a new function called `ValidateObjectParameters` (@wlynch and you suggested).
Happy to split out this further once the result side gets clear needs for another new helper function on this part. Let me know what you think. @Yongxuanzhang @ywluogg

Reply:
I will keep this in mind